### DISCUSSION TITLE
AF / regional variants

### DISCUSSION THREAD

**Discussion by lxsxl** [PE5PVB/TEF6686_ESP32#280] [Ideas]
If you still like to enhance the AF function you could implement the regional version (only if AF method B is used). This works as follows:

In AF method B for each frequency of the network a set of AFs is transmitted. A single AF-set consists of the header (= number of frequencies + lead-frequency) and a few AF-pairs, where the lead-frequency is always one part of the pair.

If the AF is the same regional version as the lead-frequency, the pair is transmitted in ascending order (e. g. lead-frequency 100,4 -> AF-pair 100,4 - 100,8 or 98,6 - 100,4),
If the AF is a regional variant to the lead-frequency, the pair is transmitted in descending order (e. g. lead-frequency 100,4 -> AF-pair 100,4 - 91,8 or 101,0 - 100,4).

I will add a screenshot of RDS Spy, this may be more illustrative. "RV" means regional variant, might also be useful for the TEF.

Be careful: One frequency can be for the same RV and also for a different one (in RDS Spy 91,8 MHz), so this would have to be shown twice.

![AF in RDS Spy](https://github.com/PE5PVB/TEF6686_ESP32/assets/140750201/c88314b1-1e02-407b-b2ee-a5acddbf844f)

**Comment by lxsxl** [comment]
And som more examples from IEC:
![AF RV from IEC](https://github.com/PE5PVB/TEF6686_ESP32/assets/140750201/2155fb4d-7058-4f2d-bed8-0215c59f2788)

**Comment by PE5PVB** [comment]
Can you get me a MPX recording on a station with AF method B?

**Comment by lxsxl** [comment]
Do you still have the one from WDR2? This should work fine. Otherwise I will record tomorrow in the evening.

**Comment by lxsxl** [comment]
One additional hint: There might be further AF that are not listed for the tuned frequency (but for other AF of the same network). If you are going to list them though, they should be marked as not belonging to the tuned frequency (but to the network), e. g. followed by "NA" or a fourth colour.

**Comment by PE5PVB** [comment]
I've added a detector for the regional flag. I can't test it here, since no stations are using it. Please test and report.

**Comment by lxsxl** [comment]
Are the "regional" AFs the blue coloured? (If yes it doesn't match actually)

**Comment by PE5PVB** [comment]
I used this lines, where buffer0 is f1 and buffer1 is f2:
                    if (buffer0 == currentfreq && buffer0 < buffer1) af[af_counter].regional = true;
                    if (buffer1 == currentfreq && buffer0 > buffer1) af[af_counter].regional = true;

**Comment by PE5PVB** [comment]
Ah, I think I found a mistake here. I've just updated it.

**Comment by andimik** [comment]
But why has the "filler" been removed? RDS Spy uses the term "RV" for *regional variant* which is easier to understand.

**Comment by PE5PVB** [comment]
I will replace the color for RV. Hopefully this will fit, otherwise I will do (r). About filler, as far as I know this is only a placeholder and does not have anything to do with the frequency. Filler is used when you have an even number of frequencies. This was misunderstood by me first and is corrected.

**Comment by PE5PVB** [comment]
Do you have any idea how to detect if a station uses method A or B?

**Comment by lxsxl** [comment]
It’s the twelveth bit in group 2 block 2 (the one between PTY and RT segment adress).Am 14.08.2023 um 17:13 schrieb Sjef Verhoeven PE5PVB ***@***.***>:﻿
Do you have any idea how to detect if a station uses method A or B?

**Comment by lxsxl** [comment]
Yes, that‘s better. So we will see both colour for reception status and RV for regional indicator. Otherwise you would need six colours.Actually I‘m too busy to test intensively, so please be patient a few days …Am 14.08.2023 um 17:10 schrieb Sjef Verhoeven PE5PVB ***@***.***>:﻿
I will replace the color for RV. Hopefully this will fit, otherwise I will do (r). About filler, as far as I know this is only a placeholder and does not have anything to do with the frequency. Filler is used when you have an even number of frequencies. This was misunderstood by me first and is corrected.

**Comment by lxsxl** [comment]
The filler is worthless as it only indicates an empty space within the AF pairs in method A (that is always when the number of AFs is even.Am 14.08.2023 um 11:48 schrieb andimik ***@***.***>:﻿
But why has the "filler" been removed? RDS Spy uses the term "RV" for regional variant which is easier to understand.

**Comment by PE5PVB** [comment]
No, AF method B is not even, because there is a leading number of AF's. For example
04, 99.5
104.5, 99.5
99.5 100.1
93.1 99.5
91.9 99.5

The 04 is, 4 AF frequencies follows. So method B should always be uneven and there is no need for a filler. In method A it can be even and then you need a filler for the unused space.

**Comment by PE5PVB** [comment]
> It’s the twelveth bit in group 2 block 2 (the one between PTY and RT segment adress)

I can't find this in the documentation. Can you show me where you found this?

**Comment by lxsxl** [comment]
> No, AF method B is not even, because there is a leading number of AF's. For example 04, 99.5 104.5, 99.5 99.5 100.1 93.1 99.5 91.9 99.5
> 
> The 04 is, 4 AF frequencies follows. So method B should always be uneven and there is no need for a filler. In method A it can be even and then you need a filler for the unused space.

That's exactly what I tried to explain. Sorry if was likely misunderstood...

**Comment by lxsxl** [comment]
> > It’s the twelveth bit in group 2 block 2 (the one between PTY and RT segment adress)
> 
> I can't find this in the documentation. Can you show me where you found this?

You can find it on pdf-page 55 (group typ 2A and 2B):
![RT Flag A-B](https://github.com/PE5PVB/TEF6686_ESP32/assets/140750201/9044747b-a5ac-43c5-b82b-264d81caf85f)

**Comment by PE5PVB** [comment]
Yes, this is the RT AB flag, but I mean a way to detect if a station uses AF method A or AF method B

**Comment by lxsxl** [comment]
Sorry for this ... Then let's go ahead with AF:

1. There is no A/B-flag for AFs. It has to be determined logically if AF follows method A or method B.
2. If it is method A, there should be no double entry as only ONE list of frequencies with max. 25 AFs is transmitted.
3. If it is method B, there are several list being transmitted. One of them is the list for the tuned frequency (TF), the others refer to other frequencies of the same network. It is characteristic for method B that the frequency in the first group 0A (the one in pair with the code for the number of frequencies) is within each following group 0A - as long as the list for this frequency is transmitted. E. g.: an AF-list with 5 AFs will need 6 0A groups (1. code for AF-number + TF // 2. AF1 + TF // 3. AF2 + TF // 4. AF3 + TF // 5. TF + AF4 // 6. TF + AF5). After that the next AF-list for a different transmitter will follow. To get the right AFs for the TF you thus have to watch for completeness of 0A-groups (at least for the TF).
4. A method B AF-list can (and shall!) be split into two lists when the number of AF is greater than 12. Those two (or even more) lists shall follow one another to indicate that they belong to each other.
5. If there are two method B AF-lists for the same TF but they are disconnected by an AF-list for different TF, those lists do not belong to each other (the same frequency is used for two different transmitters of this network).
6. If the frequencies within one pair are in ascending order, both transmit the same regional variant, if they are in descending order, they transmit regional variants.

So far the regulation of IEC. What would be the best for displaying on the TEF:

1. As long as the method is not safely detected, show all received AFs without an ID for method A or B.
2. If method A is detected, add an ID for method A.
3. If method B is detected, add an ID for method B, add the "R" for RV-AFs and add "D" (or a different character) for frequencies that are transmitted in AF-lists for other frequencies than the TF.
4. One AF can be transmitted twice in a method B AF-list, then it should also be displayed twice (once "non-R" and once "R").
5. In case of method B you could probably also add a number of AF-lists that are transmitted (but less important).

I think this would be a solution as complete as possible (and useful) for the TEF. Hope, that it is not exceeding the remaining memory...

**Comment by PE5PVB** [comment]
I can try to make an algorithm that counts the number of current frequencies between the indentifications of (x AF's will follow). In method B it should always be more than 1 time. Is that an idea?

**Comment by lxsxl** [comment]
I think this should work.

**Comment by PE5PVB** [comment]
I've added a script do detect this, please test.

**Comment by lxsxl** [comment]
Actually it still doesn‘t work. There are no RV marked for any station, and I checked a lot that do have them. At least the incorrect RV-markings have gone.
Could you implement a flag that shows If you detected method A or B, because actually I can‘t even evaluate this. And as this is the first step it could be helpful.

**Comment by andimik** [comment]
If you want to check if A or B is used, why not use RDSSPY? It will also help you to identify possible stations which are using that function.

And you could share either a 192 kHz 16bit mono mpx recording or at least a spy file with the RDS.

**Comment by PE5PVB** [comment]
Can you get me a MPX recording for testing, including the frequency from the recording?

**Comment by lxsxl** [comment]
Flag for method A/B could help to see If at least the correct method was detected by the TEF (which I compare with RDS Spy of course). German public broadcasters aImost completely use method B, so I don‘t have to compare.

**Comment by lxsxl** [comment]
***@***.*** sent you MPX and Spy recordings

1 item, 206 MB in total ・ Expires on 14 September, 2023


Download link:
https://wetransfer.com/downloads/fc7e38ab9ac259a99d6352ba927fd63620230907195816/b893453af6b3434074bb87937e3ec97c20230907195945/423ca9?trk=TRN_TDL_01&utm_campaign=TRN_TDL_01&utm_medium=email&utm_source=sendgrid





1 item

AF method B examples.zip - 206 MB


[Get more out of WeTransfer, get Pro](https://wetransfer.com/pro?trk=TRN_TDL_01&utm_campaign=TRN_TDL_01&utm_medium=email&utm_source=sendgrid)



About WeTransfer: https://wetransfer.com/about
Help: https://wetransfer.zendesk.com/hc/en-us
Legal: https://wetransfer.com/legal/terms


To make sure our emails arrive, please add ***@***.*** to [your contacts](https://wetransfer.zendesk.com/hc/en-us/articles/204909429).

**Comment by PE5PVB** [comment]
Found it. Please test again.

**Comment by lxsxl** [comment]
I will do later this evening and let you know!

**Comment by lxsxl** [comment]
Something has changed, for some AFs the "R" is visible now. But it is not reproducible: If you tune to the same frequency later the "R" may have gone. I suppose that it is due to fact that AF method B uses several AF-lists and the RV-status may differ in each list. So there is still something misinterpreted.

If you check my recordings ([https://we.tl/t-bokw2VdTBc](https://we.tl/t-bokw2VdTBc)) you can see how RDS Spy is working on AF method B: At the beginning the frequencies are added to method A-section. Step by step the single AF-lists are detected as method B and tranferred to the method B-section. I suppose that Spy first collects all AF-pairs for the list, checks if the head-frequency is in the leading pair (together with the lists AF-number) AND in each AF-pair, then it is safely method B. In that way each AF-list is checked and processed. It is important that you read the RV-status only from one single AF-list (the one that refers to the tuned frequency) because RV-status may be different in another list (i. e. for a different frequency). And also important: If the TEF switches to another AF, the AF-list has to be rebuild (then of course refering to the new tuned frequency, I don't know if this is actually already done).

Idk if such routines are possible to be done by the TEF? Maybe it is too much data-collecting and logics for it? Anyway, I will check deeper in the next days and try to find some more reasons.

**Comment by PE5PVB** [comment]
What I currently do is this:
1. Wait for the AF index (the number of AF's to follow).
2. Count the times the current frequency is read from the AF A or AF B.
3. If number of times current frequency was there exceeds 2, then Method B flag is set high

Please note. When opening AF view and all AF frequencies are read, but the flag was not set, the page will not be refreshed automaticly. I have to add this. So currently, just tune into a station, wait a while and then open AF to see the regional idents.

**Comment by lxsxl** [comment]
AF RV still don't work most of the time. Only randomly some frequencies are detected as RV, and if so, the R-markings are incorrect most cases. But as I often cannot see any R-markings, I wonder if at least method B is set high (so if you could implement a hint on that, e. g. two dots alternating in red and green, similar to the description in RT A/B-flag) we could at least see if this works fine).
What might lead to problems in detecting method A/B and/or RV could be this:
1. Do you concentrate on the AF-list that is transmitted for the current frequency? If you compare to a different AF-list than this one it may happen that you don't receive the current frequency at all.
2. If you extracted/collected the specific AF-list for the current frequency, you can evaluate with the code you described above:
(if (buffer0 == currentfreq && buffer0 < buffer1) af[af_counter].regional = true;
if (buffer1 == currentfreq && buffer0 > buffer1) af[af_counter].regional = true; [shouldn't this be = false? sorry, I'm not familiar with programming...]
3. If you want to display also the AFs that are not within the specific list for the current frequency you can add them to the TEF-list but better with an extra-marking (e. g. "F" for "further AF") as you are not able to detect if it is an RV or not.

Maybe the following screenshots from RDS Spy may help, each specific AF-list is transmitted on the same frequency. So you see, if you refer to the 93,2-list there is no 100,4 (or others) at all, and so I suppose method B probably won't be detected:
![AF RV SPY-samples](https://github.com/PE5PVB/TEF6686_ESP32/assets/140750201/6ba88f85-07e8-4795-8402-12dfa90ff247)

**Comment by PE5PVB** [comment]
Well, to do it correct, we have to build a complete database and compare everything which is a lot of work. Maybe this is something for the future, or, if someone want to add some time in it, please do.

**Reply by lxsxl** [reply]
Yes, I totally agree. That‘s why I asked already if this developement may have been one step too big. So from my point of view feel free to head back to the state without RV to avoid wrong interpretations.
Or at least one question: in the new Beta-release you mentioned a fix for AF method B. What is it about? I will only be able to test again from 9th of october…

**Comment by PE5PVB** [comment]
Did you check with the latest build? I did some small improvements which can affect this issue.

**Comment by PE5PVB** [comment]
I close this one, because futher things are issues, and should be put there.
